### PR TITLE
[0.2.1] Simple multi threading client-server implementation + start to logging with log4j + small refactoring in ship-placement package

### PR DESCRIPTION
**Pull request by totine**
It will be improved

### CODE REVIEW

**Review by boczkas — CHANGES_REQUESTED** [review]
mci doesn't work:
[ERROR] /home/boczkas/IdeaProjects/recki/Asia/OStatki/src/main/java/placement/model/ship/DirectedShip.java:40: Line is longer than 120 characters (found 137). [LineLength]
Audit done.
[INFO] There is 1 error reported by Checkstyle 6.18 with checkstyle.xml ruleset.
[ERROR] src/main/java/placement/model/ship/DirectedShip.java:[40] (sizes) LineLength: Line is longer than 120 characters (found 137).
[INFO] ------------------------------------------------------------------------

File: client/src/main/java/connection/ClientIO.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+/**
+ * Wrapper for client socket input and output
+ */
+class ClientIO {
+    private final PrintWriter out;
+    private final Scanner in;
+
+    private ClientIO(Socket socket) throws IOException {
+        OutputStream outputStream = socket.getOutputStream();
+        OutputStreamWriter outputStreamWriterUTF8 = new OutputStreamWriter(outputStream, StandardCharsets.UTF_8);
```
Comment (boczkas) [inline client/src/main/java/connection/ClientIO.java]: will you have different writer than UTF8?
If not name can be changed

File: client/src/main/java/connection/ClientIO.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        this.out = new PrintWriter(outputStreamWriterUTF8, true);
```
Comment (boczkas) [inline client/src/main/java/connection/ClientIO.java]: true can be extracted to:
boolean FLUSH = true;

File: server/src/main/java/MessagesFromClientHandler.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import java.io.IOException;
```
Comment (boczkas) [inline server/src/main/java/MessagesFromClientHandler.java]: Server packages?

File: server/src/main/java/MessagesToClientHandler.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+/**
+ * encapsulates client's side in and out streams
+ */
+class MessagesToClientHandler {
+    private PrintWriter out;
+
+    MessagesToClientHandler(Socket clientSocket) throws IOException {
+        OutputStream outputStream = clientSocket.getOutputStream();
+        OutputStreamWriter outputStreamWriterUTF8 = new OutputStreamWriter(outputStream, StandardCharsets.UTF_8);
+        this.out = new PrintWriter(outputStreamWriterUTF8, true);
```
Comment (boczkas) [inline server/src/main/java/MessagesToClientHandler.java]: new in constructor?

File: server/src/main/java/MessagesToClientHandler.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+import java.io.OutputStreamWriter;
+import java.io.PrintWriter;
+import java.net.Socket;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * encapsulates client's side in and out streams
+ */
+class MessagesToClientHandler {
+    private PrintWriter out;
```
Comment (boczkas) [inline server/src/main/java/MessagesToClientHandler.java]: final?

File: server/src/main/java/Server.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+ * to listen for client connections.
+ */
+class Server {
+    private final ServerSocket serverSocket;
+    private static final Logger LOGGER = LogManager.getLogger(Server.class);
+    private List<MessagesToClientHandler> clientSocketList;
+
+    private Server(ServerSocket serverSocket) {
+        this.serverSocket = serverSocket;
+        this.clientSocketList = new ArrayList<>();
```
Comment (boczkas) [inline server/src/main/java/Server.java]: new in constructor

File: server/src/main/java/Server.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.List;
+
+/**
+ * Class for server object. When calling start(), it starts
+ * to listen for client connections.
+ */
+class Server {
+    private final ServerSocket serverSocket;
+    private static final Logger LOGGER = LogManager.getLogger(Server.class);
+    private List<MessagesToClientHandler> clientSocketList;
```
Comment (boczkas) [inline server/src/main/java/Server.java]: doesn't it deserve to separate class?

File: server/src/main/java/Server.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+    private Server(ServerSocket serverSocket) {
+        this.serverSocket = serverSocket;
+        this.clientSocketList = new ArrayList<>();
+    }
+
+    static Server createServer(int serverPort) {
+        ServerSocket serverSocket = null;
```
Comment (boczkas) [inline server/src/main/java/Server.java]: null?

File: server/src/main/java/Server.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Class for server object. When calling start(), it starts
+ * to listen for client connections.
+ */
+class Server {
+    private final ServerSocket serverSocket;
+    private static final Logger LOGGER = LogManager.getLogger(Server.class);
```
Comment (boczkas) [inline server/src/main/java/Server.java]: LOGGER with capital letters suggest that it is constant, not an object

File: server/src/main/java/Server.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        }
+        return new Server(serverSocket);
+    }
+
+    void start() {
+        while (true) {
+            try {
+                Socket clientSocket = serverSocket.accept();
+                LOGGER.info(clientSocket.toString() + " connected");
+                MessagesFromClientHandler clientIn = new MessagesFromClientHandler(clientSocket, this);
```
Comment (boczkas) [inline server/src/main/java/Server.java]: You have nice names for MessagesFrom/ToClientHandler why have you named it's object not so nice - like clientIn?
I have a feeling that it is mixed of abstraction levels

File: shared/src/main/java/placement/controller/RandomFleetPlacer.java [hunk 11]
```diff
@@ -22,12 +22,15 @@
     private final Random random;
 
 
-    RandomFleetPlacer(Fleet<UndirectedShip> fleet, Board board) {
+    private RandomFleetPlacer(Fleet<UndirectedShip> fleet, Board board) {
         this.fleet = fleet;
         this.board = board;
         this.random = new Random();
```
Comment (boczkas) [inline shared/src/main/java/placement/controller/RandomFleetPlacer.java]: new in constructor

**Review by boczkas — CHANGES_REQUESTED** [review]
remove new from constructors

File: client/src/main/java/connection/ConnectionAttempt.java [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+        try {
+            String ip = "localhost";
+            client = ClientIO.createClient(ip, PORT_NUMBER);
+            Scanner scanner = new Scanner(System.in, String.valueOf(StandardCharsets.UTF_8));
+            new Thread(client::getMessage).start();
+            while (scanner.hasNextLine()) {
+                client.sendMessage(scanner.nextLine());
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (krzysieksr) [inline client/src/main/java/connection/ConnectionAttempt.java:27]: Please add some external logger for the future.

File: checkstyle.xml [hunk 13]
```diff
@@ -165,7 +162,7 @@
 
         <!-- Checks for class design                         -->
         <!-- See http://checkstyle.sf.net/config_design.html -->
-        <module name="FinalClass"/>
+        <!--<module name="FinalClass"/>-->
```
Comment (krzysieksr) [inline checkstyle.xml:165]: Remove commented modules or make them dissabled.

File: client/src/main/java/gui/GUIStarter.java [hunk 14]
```diff
@@ -24,10 +24,11 @@ public final void start(Stage primaryStage) throws Exception {
     }
 
     /**
-     * main method - actual entry point of the application
+     * chat method - actual entry point of the application
      * @param args - console arguments used by FX
      */
     public static void main(String[] args) {
-        launch(args);
+
```
Comment (krzysieksr) [inline client/src/main/java/gui/GUIStarter.java]: Are you sure, that main method should be empty?

**Review by krzysieksr — APPROVED** [review]
Make little refactor and you can merge it

File: client/src/main/java/connection/ClientIO.java [hunk 15]
```diff
@@ -17,8 +17,9 @@
 
     private ClientIO(Socket socket) throws IOException {
         OutputStream outputStream = socket.getOutputStream();
-        OutputStreamWriter outputStreamWriterUTF8 = new OutputStreamWriter(outputStream, StandardCharsets.UTF_8);
-        this.out = new PrintWriter(outputStreamWriterUTF8, true);
+        OutputStreamWriter outputStreamWriter = new OutputStreamWriter(outputStream, StandardCharsets.UTF_8);
+        boolean autoFlush = true;
+        this.out = new PrintWriter(outputStreamWriter, autoFlush);
```
Comment (boczkas) [inline client/src/main/java/connection/ClientIO.java]: new in constructor

File: client/src/main/java/connection/ClientIO.java (hunk 2 above)
Comment (boczkas) [reply client/src/main/java/connection/ClientIO.java]: Still not extracted

File: server/src/main/java/server/Server.java [hunk 16]
```diff
@@ -1,46 +1,55 @@
...
+public class Server {
     private final ServerSocket serverSocket;
     private static final Logger LOGGER = LogManager.getLogger(Server.class);
     private List<MessagesToClientHandler> clientSocketList;
 
-    private Server(ServerSocket serverSocket) {
+    private Server(ServerSocket serverSocket, List<MessagesToClientHandler> list) {
         this.serverSocket = serverSocket;
-        this.clientSocketList = new ArrayList<>();
+        this.clientSocketList = list;
```
Comment (boczkas) [inline server/src/main/java/server/Server.java]: list - use better name

File: server/src/main/java/server/Server.java [hunk 17]
```diff
@@ -1,46 +1,55 @@
...
                 Socket clientSocket = serverSocket.accept();
                 LOGGER.info(clientSocket.toString() + " connected");
-                MessagesFromClientHandler clientIn = new MessagesFromClientHandler(clientSocket, this);
-                MessagesToClientHandler clientOut = new MessagesToClientHandler(clientSocket);
-                clientSocketList.add(clientOut);
-                new Thread(clientIn).start();
-
+                MessagesFromClientHandler messagesFromClientHandler = new MessagesFromClientHandler(clientSocket, this);
+                MessagesToClientHandler messagesToClientHandler = new MessagesToClientHandler(clientSocket,
+                        new PrintWriter(
```
Comment (boczkas) [inline server/src/main/java/server/Server.java]: extract variables from new

**Review by boczkas — APPROVED** [review]
:+1:

**Pull request merged**